### PR TITLE
Add a build service for building Rust projects through Cargo

### PR DESCRIPTION
**Pull request by skade**
/cc @ashleygwilliams 

This rebuilds the rustc endpoint to use cargo instead. Instead of `{ opts: {...}, src: ...}` it now accepts `{ opts {...}, tar: "..." }` where `tar` is a base64 encoded tar archive of a cargo project. The project will be untarred into a subdirectory of the temp directory and then built.

Tar archives are safe against size-based attacks and can be easily produced in the browser.

The resulting wasm file is found using the build-plan.

Note that this options the option of using wasm-bindgen through a dependency.

This PR adds an example to try out the service.

Notably missing: cleanup. How do I properly delete a directory recursively in vanilla node?

### CODE REVIEW

**Comment by skade** [comment]
Currently outstanding issues:

* [x] `cc(1)` is not being found with wasm_bindgen example
* [x] buildplan parsing does not take complex buildplans with dependencies into account

**Comment by alexcrichton** [comment]
One of the downsides of Cargo-based compilation is that when you can pull in dependencies you can, well, pull in a lot of dependencies! In that sense it may be prudent to do a few things maybe? (not necessarily all)

* Add a whitelist of crates allowed to download and build
* Avoid updating the registry on every build
* Have a time limit for the compilation process
* Somehow cache dependencies between builds (like wasm-bindgen)

Unfortunately all of those are pretty tricky :(

**Review by yurydelendik — CHANGES_REQUESTED** [review]
* [ ] existing rustc end point needs to be kept without any change
* [ ] add cargo end point
* [ ] handling of tar adds unneeded overhead (on client side mostly) -- send a dictionary of files as is
* [ ] sanitize all options in the e.g. plan["invocations"][0]["env"]

**Comment by skade** [comment]
@alexcrichton I'm aware of this, but the PR comes out of the wish to present a proper environment this weekend. Otherwise, we'd have to present a single file with a lot of gory details and have a lot of "ignore this"-explaining to do. What we can to is not accept a Cargo.toml and provide our own. Caching is mostly a problem due to the platform chosen, if I see right?

@yurydelendik I'll add the rustc endpoint back in.

I don't agree that tar requires much effort. https://github.com/beatgammit/tar-js is as easy to use as any custom built file-pack-as-json-abstraction and easier to use on both sides.

plan options are completely read by cargo locally.

**Comment by yurydelendik** [comment]
> I don't agree that tar requires much effort. https://github.com/beatgammit/tar-js is as easy to use as any custom built file-pack-as-json-abstraction and easier to use on both sides.

I cannot find tar reading capabilities in the library above -- the cargo endpoint needs to sanitize filenames.

**Comment by skade** [comment]
Your comment referred specifically to client side capabilities. There's tar libraries for the server side node, but shelling out to command line tar is feasible in this case. `tar(1)` also armors against traversal down, so it already sanitises.

I don't see how filename santisation would not also apply to any JSON based format.

**Comment by skade** [comment]
@alexcrichton I changed the current implementation so that it rejects build plans with more then 1 invocation. That could be extended to more complex rules.

On registry/cargo issues: wouldn't it be possible to make sure the slug uses one target directory for all and supply it with WASM-bindgen already built?

Also, notably missing:
* [x] `cc(1)` is not available within the slug on Heroku

File: app/src/cargo.js [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+  tempDir,
+  wasmBindgenCmd,
+  wasmBindgenDeps,
+} = require("../config.js");
+const { exec, joinCmd, exists, writeFile, readFile, mkdir, unlink } = require("./common.js");
+
+function checkBuildPlan(plan) {
+  let success = true;
+  let invocations = plan["invocations"];
+  console.log(invocations.length);
```
Comment (tschneidereit) [inline app/src/cargo.js]: Please remove these debugging log messages, here and below.

File: app/src/cargo.js [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+        ret.wasmBindgenJs = (await readFile(baseName + '.js')).toString();
+      } else {
+        await exec(joinCmd([wasmGCCmd, wasmFile]));
+        wasm = await readFile(wasmFile);
+      }
+      ret.output = wasm.toString('base64');
+      return ret;
+    } finally {
+      if (success) {}
+        //await unlink(wasmFile);
```
Comment (tschneidereit) [inline app/src/cargo.js:106]: Is this something we don't need after all? If so, please remove (and below). If yes, but not immediately, can you add a "TODO"?

**Comment by skade** [comment]
So, just to sum up the current state of things:
* Adds a `/cargo` endpoint that builds using cargo instead
* It uses cargo buildplans to validate the Cargo.toml
  - Additionally, it applies a rule to the buildplan, which is currently "only one build step" (effectively: no dependencies. The endpoint returns a proper error message.
  - That also means: currently no `#[wasm_bindgen]` annotations
* The service accepts a base64 -encoded `tar` file, which is extracts using `tar(1)` for security reasons
  - Such a file can be generated on the frontend using `tar-js` easily

Missing, but possible as extensions later:
* Allow a limited set of dependencies
* Prebuild a couple of important libraries, such as wasm-bindgen.

**Review by tschneidereit — CHANGES_REQUESTED** [review]
@skade, thank you so much for working on this! I left two tiny comments in the review, otherwise I think the code looks great.

To expand a bit on the concerns @alexcrichton and @yurydelendik voiced, we had expected something much less flexible than this - essentially the ability to throw multiple `.rs` files into a directory and invoke them in the same environment with otherwise unchanged parameters. Long-term it is much nicer to have a more flexible setup, so this is great. The downside of providing a much more powerful and efficient system right now is that we have to do more work ensuring that we don't open any holes that'd allow arbitrary code execution on the server - and that we have to do more changes on the frontend than expected.

To be clear, the way you set this up makes a lot of sense, and we should've communicated better how restricted a solution we envisioned - my apologies for that.

Now, to get this landed I think we do need to take out the `Cargo.toml` file and instead use a hard-coded one on the server. Given that we now have the `rustc` endpoint preserved, we should then be able to land and deploy this, and then work on the frontend side of it. I think I see how to use the `tar-js` library, and including that in wasm studio doesn't seem too hard indeed.

**Review by yurydelendik — APPROVED** [review]

**Review by tschneidereit — APPROVED** [review]

**Comment by yurydelendik** [comment]
Thank you for the patch

**Pull request merged**